### PR TITLE
DropdownMenu V2 tweaks

### LINKED ISSUE #55933: DropdownMenu v2: Design tweaks
https://github.com/WordPress/gutenberg/assets/846565/7ebb46f7-8edd-47ea-a1f4-df66f03e7ab0

Some design details may be subject to change, but the full spec for updated MenuItems/MenuItemFlyouts can be found in Figma [here](https://www.figma.com/file/oituj8yH1O6z1Y6EsHY6FA/DropdownMenu?type=design&node-id=586%3A3540&mode=design&t=XJ4sxwmuUmjr15FG-1).

The prototype in the video above can be tried [here](https://www.figma.com/proto/oituj8yH1O6z1Y6EsHY6FA/DropdownMenu?page-id=611%3A8843&type=design&node-id=611-7403&viewport=2046%2C295%2C0.3&t=Oq0tQjRNiqukG7TK-1&scaling=min-zoom&starting-point-node-id=611%3A7403&mode=design).

### List of tweaks

<details>
<summary>Click to expand</summary>

- [x] Remove group label
- [x] Expose Help Text subcomponent, including truncation
- [x] Allow nested triggers to have a suffix while preserving the chevron
- [x] Tweak submenus overlap
- [x] Add auto-indent functionality 
    - [x] ~Using context~
    - [x] Using CSS grid
- [x] When suffix is not present, children fill all available space
- [x] Add more examples to Storybook:
    - [x] Indented / unintended
- [x] Review sizes:
    - [x] label font size
    - [x] help text font size
    - [x] chevron size
    - [x] check and radio size
    - [x] icon size
- [x] Review spacing:
    - [x] menu padding
    - [x] item spacing
    - [x] divider spacing
    - [x] min and max menu widths
    - [x] prefix spacing
    - [x] suffix spacing
    - [x] chevron spacing
- [x] Review all text/bg colors on
  - [x] Label:
    - [x] resting
    - [x] hover
    - [x] focus
    - [x] disabled
    - [x] active flyout
  - [x] Help text
    - [x] resting
    - [x] hover
    - [x] focus
    - [x] disabled
    - [x] active flyout
  - [x] Prefix
    - [x] resting
    - [x] hover
    - [x] focus
    - [x] disabled
    - [x] active flyout
  - [x] Suffix
    - [x] resting
    - [x] hover
    - [x] focus
    - [x] disabled
    - [x] active flyout
  - [x] Chevron
    - [x] resting
    - [x] hover
    - [x] focus
    - [x] disabled
    - [x] active flyout
  - [x] Bg color
    - [x] resting
    - [x] hover
    - [x] focus
    - [x] disabled
    - [x] active flyout
- [x] Review border and separator color, including drop shadow
- [x] Truncate label text

</details>

### Questions:
- are there reusable line-height variables?
- Box shadows: spec vs existing popover box shadow?
- border colors: spec vs existing variables?
- toolbar variable?

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Closes #55933

This PR includes a series of tweaks to the experimental new version of `DropdownMenu`, based on the specs shared in #55933.

Main tweaks:
- Removed group label component (not needed)
- Added new subcomponents for menu item label and help text (mostly needed because of text truncation behavior)
- Added auto-indentation of menu items based on the presence, in the same popover, of at least one item with a prefix
- Tweaks to spacing, typography, colors, hovered/focused/disabled states...

## Why?

The design spec brings a coherent look with the rest of the components library, and addresses a few complex scenarios when dealing with some combinations of menu items and their contents.

## How?

The most complex bits in the implementation:
- CSS grid and subgrid were used to implement the auto-indent functionality
- The Label and Help text subcomponent use the `Truncate` component under the hood

## Testing Instructions

Open the "DropdownMenu v2 ariakit" Storybook examples, make sure that the component follows the design specs

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/1083581/398c4224-83f2-4219-b14b-3d739630c75c

### CODE REVIEW

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 1]
```diff
@@ -21,12 +22,15 @@ const ANIMATION_PARAMS = {
...
-
-// TODO: should bring this into the config, and make themeable
-const DEFAULT_BORDER_COLOR = COLORS.ui.borderDisabled;
+const ITEM_PADDING_BLOCK = space( 2 );
+const ITEM_PADDING_INLINE = space( 3 );
+
+// TODO:
+// - those values are different from saved variables?
+// - should bring this into the config, and make themeable
+// - border color and divider color are different?
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts:31]: Flagging this for us to discuss

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 2]
```diff
@@ -71,7 +75,13 @@ export const DropdownMenu = styled( Ariakit.Menu )<
 	/* TODO: is there a way to read the sass variable? */
 	z-index: 1000000;
 
-	min-width: 220px;
+	display: grid;
+	grid-template-columns: minmax( 0, max-content ) 1fr;
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts:79]: This is where the main grid layout is defined.

The first column is for the prefix (and can collapse to 0 width if no prefixes are rendered within the same dropdownmenu popover).

The second column take all remaining available space, and is meant to host the wrapper element containing label, help text, and suffix.

File: packages/components/src/dropdown-menu-v2-ariakit/index.tsx [hunk 3]
```diff
@@ -58,13 +57,19 @@ export const DropdownMenuItem = forwardRef<
...
 			store={ dropdownMenuContext?.store }
 		>
-			{ prefix && (
-				<Styled.ItemPrefixWrapper>{ prefix }</Styled.ItemPrefixWrapper>
-			) }
-			{ children }
-			{ suffix && (
-				<Styled.ItemSuffixWrapper>{ suffix }</Styled.ItemSuffixWrapper>
-			) }
+			<Styled.ItemPrefixWrapper>{ prefix }</Styled.ItemPrefixWrapper>
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/index.tsx:61]: The prefix wrapper is always rendered (even when empty) in order to correctly render in the first column of the grid layout.

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 4]
```diff
@@ -110,102 +120,47 @@ export const DropdownMenu = styled( Ariakit.Menu )<
...
 
-export const ItemSuffixWrapper = styled.span`
-	${ itemSuffix }
-`;
+	/*
+	 * Define a grid layout which inherits the same columns configuration
+	 * from the parent layout (ie. subgrid).
+	 */
+	display: grid;
+	grid-template-columns: subgrid;
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts:138]: CSS subgrid enables us to define a grid layout at the menu level, and "propagate" it down all the way to the menu item children.

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 5]
```diff
@@ -224,74 +179,151 @@ const baseItem = css`
...
-	/* TODO: font size doesn't match available ones via "font" utils */
-	font-size: 11px;
-	line-height: 1.4;
-	font-weight: 500;
-	text-transform: uppercase;
+	justify-content: center;
+
+	/* Override inline styles applied by ariakit */
+	width: auto !important;
+	height: auto !important;
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts]: @diegohaz I wonder if this is hacky and/or risky? Ariakit seems to be adding inline styles for `width: 1em` and `height: 1em`

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 6]
```diff
@@ -224,74 +179,151 @@ const baseItem = css`
...
+`;
+
+export const DropdownMenuItemContentWrapper = styled.div`
+	/*
+	 * Always occupy the second column, since the first column
+	 * is taken by the prefix wrapper (when displayed).
+	 */
+	grid-column: 2;
+
+	display: flex;
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts:243]: This wrapper occupies the second "column" of the grid layout, and is a flexbox container. This mixed grid + flexbox setup enables menu items to align their left indentation based on the presence of at least one prefix, but keeps independent around the right-side alignment

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 7]
```diff
@@ -224,74 +179,151 @@ const baseItem = css`
...
+	[data-active-item]:not( [data-focus-visible] ) *:not(${ DropdownMenu }) &,
+	/* When the parent menu item is disabled */
+	[aria-disabled='true'] *:not(${ DropdownMenu }) & {
+		color: inherit;
+	}
+`;
+
+export const DropdownMenuGroup = styled( Ariakit.MenuGroup )`
+	/* Ignore this element when calculating the layout. Useful for subgrid */
+	display: contents;
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts:283]: Display contents allows us to ignore this element when calculating the layout, and look at its children directly

File: packages/components/src/dropdown-menu-v2-ariakit/index.tsx [hunk 8]
```diff
@@ -280,12 +288,16 @@ const UnconnectedDropdownMenu = (
...
-										size={ 24 }
-									/>
+								suffix: (
+									<>
+										{ trigger.props.suffix }
+										<Styled.SubmenuChevronIcon
+											aria-hidden="true"
+											icon={ chevronRightSmall }
+											size={ 24 }
+											preserveAspectRatio="xMidYMid slice"
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/index.tsx:305]: `preserveAspectRatio` is used to resize the width of the icon while preserving its real size (ie. slicing the SVG viewport)

File: packages/components/src/dropdown-menu-v2-ariakit/stories/index.story.tsx [hunk 9]
```diff
@@ -383,7 +450,13 @@ const Slot = () => {
 		[ dropdownMenuContext ]
 	);
 
-	return <ExampleSlotFill.Slot fillProps={ fillProps } bubblesVirtually />;
+	return (
+		<ExampleSlotFill.Slot
+			fillProps={ fillProps }
+			bubblesVirtually
+			style={ { display: 'contents' } }
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/stories/index.story.tsx:457]: Adding `display: contents` to the `Slot` element is necessary, otherwise the slotfill example breaks the grid layout.

File: packages/components/src/slot-fill/types.ts
Comment (ciampo) [inline packages/components/src/slot-fill/types.ts:1]: These changes should be likely moved to a separate PR

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts [hunk 10]
```diff
@@ -224,74 +179,151 @@ const baseItem = css`
...
-			transform: `scaleX(-1) translateX(${ space( 2 ) })`,
+			transform: `scaleX(-1)`,
 		}
-	) }
+	) };
+`;
+
+export const DropdownMenuItemLabel = styled( Truncate )`
+	font-size: ${ font( 'default.fontSize' ) };
+	line-height: 20px;
```
Comment (ciampo) [inline packages/components/src/dropdown-menu-v2-ariakit/styles.ts:320]: Do we have any presets for line-height? Or are these values ad-hoc for this component?

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 5 above)
Comment (diegohaz) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts]: @ciampo I don't think it's risky. But if you want something less hacky, you can pass the inline styles directly to the component:

```jsx
<MenuItemCheck style={{ width: "auto", height: "auto" }}>
```

**Review by chad1008 — COMMENTED** [review]
This all looks good and tests well to me - the only deviation from the spec that jumped out to me was that disabled items aren't getting a `not-allowed` cursor when hovered, but that may be something you're planning to address separately.

**Comment by andrewhayward** [comment]
I don't know if this is the right place for this feedback, but for want of anywhere else...

### Scroll margins

When there isn't enough space to render the whole menu, it gets a scrollbar, which is fine, but the menu items need a scroll margin to make sure they're fully visible.

A menu with a scrollbar. The first menu item has a focus ring, cut off by the scroll position.

```diff
--- a/packages/components/src/dropdown-menu-v2-ariakit/styles.ts
+++ b/packages/components/src/dropdown-menu-v2-ariakit/styles.ts
@@ -188,16 +188,21 @@ const baseItem = css`
        }
 `;

-export const DropdownMenuItem = styled( Ariakit.MenuItem )`
+const menuItem = css`
        ${ baseItem };
+       scroll-margin: ${ CONTENT_WRAPPER_PADDING };
+`;
+
+export const DropdownMenuItem = styled( Ariakit.MenuItem )`
+       ${ menuItem };
 `;

 export const DropdownMenuCheckboxItem = styled( Ariakit.MenuItemCheckbox )`
-       ${ baseItem };
+       ${ menuItem };
 `;

 export const DropdownMenuRadioItem = styled( Ariakit.MenuItemRadio )`
-       ${ baseItem };
+       ${ menuItem };
 `;

 export const ItemPrefixWrapper = styled.span`
```

### Disabled menu items

Every menu item should really be focusable, [which isn't the default with Ariakit](https://ariakit.org/reference/menu-item#accessiblewhendisabled). We should set `accessibleWhenDisabled` as standard, so that consumers have to explicitly override it.

A menu with a disabled menu item showing a focus ring.

```diff
index aac8a8844f..a318d295a8 100644
--- a/packages/components/src/dropdown-menu-v2-ariakit/index.tsx
+++ b/packages/components/src/dropdown-menu-v2-ariakit/index.tsx
@@ -53,6 +53,7 @@ export const DropdownMenuItem = forwardRef<
        return (
                <Styled.DropdownMenuItem
                        ref={ ref }
+                       accessibleWhenDisabled
                        { ...props }
                        hideOnClick={ hideOnClick }
                        store={ dropdownMenuContext?.store }
@@ -86,6 +87,7 @@ export const DropdownMenuCheckboxItem = forwardRef<
        return (
                <Styled.DropdownMenuCheckboxItem
                        ref={ ref }
+                       accessibleWhenDisabled
                        { ...props }
                        hideOnClick={ hideOnClick }
                        store={ dropdownMenuContext?.store }
@@ -130,6 +132,7 @@ export const DropdownMenuRadioItem = forwardRef<
        return (
                <Styled.DropdownMenuRadioItem
                        ref={ ref }
+                       accessibleWhenDisabled
                        { ...props }
                        hideOnClick={ hideOnClick }
                        store={ dropdownMenuContext?.store }
```

**Comment by ciampo** [comment]
Thank you @chad1008 and @andrewhayward for your reviews!

I've pushed https://github.com/WordPress/gutenberg/pull/56041/commits/310eb1219bd58452f5a9bc459f50e25327bfaa87, https://github.com/WordPress/gutenberg/pull/56041/commits/da57e59c42fd6ab37f52cdb585a5e06c86cf2703 and https://github.com/WordPress/gutenberg/pull/56041/commits/264b4fcae90568dc8ace2ddbfe5daa9efbebb632 which should address your feedback.

@jameskoster , would you be able to take a look at this PR? I also left a couple of comments ([one](https://github.com/WordPress/gutenberg/pull/56041#discussion_r1393322605), [two](https://github.com/WordPress/gutenberg/pull/56041#discussion_r1393366273)) highlighting a few points that I wanted to discuss.

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 5 above)
Comment (ciampo) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts]: https://github.com/WordPress/gutenberg/pull/56041/commits/495a29fa77be0e87a922d06742d3be2f8a3805eb

File: packages/components/src/slot-fill/types.ts
Comment (ciampo) [reply packages/components/src/slot-fill/types.ts:1]: https://github.com/WordPress/gutenberg/pull/56428

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 1 above)
Comment (jameskoster) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts:31]: > border color and divider color are different?

Yup that's fine.

> should bring this into the config, and make themeable

I'm not sure what is entailed, but the components should eventually be themable using the new color scales:

theme

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 10 above)
Comment (jameskoster) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts:320]: We have text styles in Figma, so these values aren't ad hoc. I don't know how well those styles are reflected in the codebase though. All I'm aware of is https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/text/styles/text-mixins.native.js. 

Typography is due a bit of a revamp for the admin work (https://github.com/WordPress/gutenberg/issues/53322).

**Comment by jameskoster** [comment]
@ciampo this is looking fantastic! Seeing it 'in the flesh' I do wonder if the spacing is a little large on the groups. Would it be easy to try 4px padding instead of 8, with a 4px radius on the container too?

Screenshot 2023-11-22 at 16 05 00

Edit: seems this was pretty easy to do so I pushed a commit.

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 10 above)
Comment (ciampo) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts:320]: Sounds good, then I guess we can leave this as-is and make sure we revisit after we have some better common typography styles.

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 1 above)
Comment (ciampo) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts:31]: Thank you for clarifying. 

There is one more question from the list above that I wanted to clarify — from the design spec, it seems that the border color is gray 300, and the divider is gray 200. 

While these values seem to align with the inline comments [where the grays are defined](https://github.com/WordPress/gutenberg/blob/dccbb4646c07d1ff748f3114b63ac2a4adcda84c/packages/components/src/utils/colors-values.js#L17-L20), I noticed that we have [some variables specific to borders that are not using the same colors](https://github.com/WordPress/gutenberg/blob/dccbb4646c07d1ff748f3114b63ac2a4adcda84c/packages/components/src/utils/colors-values.js#L42-L45) — I guess it would be good if you could review where that are being used, and potentially give feedback around whether we can tweak those variables to use gray 300 and 200.

**Comment by ciampo** [comment]
@jameskoster and @WordPress/gutenberg-components , I believe that this PR is ready for a last round of review. It would be great if we could merge this PR soon, since it unblocks a bunch of other dropdownmenu-related work.

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 1 above)
Comment (jameskoster) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts:31]: The second group of vars seem to be applied to the experimental `ToggleGroupControl`. They make sense in the context of an interactive element like a form input where there are stricter a11y requirements for contrast. Visual elements (like the menu container) can use the lighter borders.

**Review by jameskoster — COMMENTED** [review]
There's one tiny detail to fix. We need to update the offset position values accounting for the new padding. Notice how the flyouts aren't lining up nicely now:

Screenshot 2023-11-23 at 14 18 59

File: packages/components/src/dropdown-menu-v2-ariakit/styles.ts (hunk 1 above)
Comment (ciampo) [reply packages/components/src/dropdown-menu-v2-ariakit/styles.ts:31]: Good to know, and likely something that we should codify when working on the wider theme / design system

**Comment by ciampo** [comment]
@jameskoster thank you for the detailed screenshot! [Done](https://github.com/WordPress/gutenberg/pull/56041/commits/c0f12c343e84c05474a356213ef1f67168ef1099)

**Review by jameskoster — APPROVED** [review]
I think we're looking good. There might be some minor visual tweaks down the road as we make more practical use of the component, but those are easy to revisit in smaller PRs as required.
